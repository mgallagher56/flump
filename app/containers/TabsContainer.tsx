import type { FC, PropsWithChildren } from 'react';
import { useId } from 'react';

import { normalizeProps, useMachine } from '@zag-js/react';
import * as tabs from '@zag-js/tabs';

import FLPTab from '../components/core/tabs/FLPTab';
import FLPTabList from '../components/core/tabs/FLPTabList';
import FLPTabPanel from '../components/core/tabs/FLPTabPanel';
import FLPTabPanels from '../components/core/tabs/FLPTabPanels';
import FLPTabsWrapper from '../components/core/tabs/FLPTabs';
import type { TabData } from '../components/core/tabs/types';

interface TabsContainerProps {
  data: TabData[];
  orientation?: 'horizontal' | 'vertical';
}

const TabsContainer: FC<PropsWithChildren<TabsContainerProps>> = ({ data, orientation }) => {
  const FLPOrientation = orientation ?? 'horizontal';
  const [state, send] = useMachine(tabs.machine({ id: useId(), value: data[0].value, orientation: FLPOrientation }));

  const api = tabs.connect(state, send, normalizeProps);

  return (
    <FLPTabsWrapper {...api.rootProps} orientation={FLPOrientation}>
      <FLPTabList {...api.tablistProps}>
        {data.map((item, index) => (
          <FLPTab {...api.getTriggerProps({ value: item.value })} tabIndex={index} key={item.value}>
            {item.label}
          </FLPTab>
        ))}
        <div {...api.indicatorProps} />
      </FLPTabList>
      <FLPTabPanels>
        {data.map((item, index) => (
          <FLPTabPanel {...api.getContentProps({ value: item.value })} tabIndex={index} key={item.value}>
            {item.content}
          </FLPTabPanel>
        ))}
      </FLPTabPanels>
    </FLPTabsWrapper>
  );
};

export default TabsContainer;
