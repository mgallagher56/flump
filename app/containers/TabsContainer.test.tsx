import { render, screen } from '@testing-library/react';
import { describe, expect, test } from 'vitest';
import type { TabData } from '~/components/core/tabs/types';

import TabsContainer from './TabsContainer';

const mockTabData: TabData[] = [
  {
    content: <div>Tab content 1</div>,
    isDisabled: false,
    label: 'Tab 1',
    value: 'tab1'
  },
  {
    content: <div>Tab content 2</div>,
    isDisabled: false,
    label: 'Tab 2',
    value: 'tab2'
  },
  {
    content: <div>Tab content 3</div>,
    isDisabled: true,
    label: 'Tab 3',
    value: 'tab3'
  }
];

const ariaSelectedValue = (tab: HTMLElement): string => tab['_attributes']['aria-selected'].value;

describe('TabsContainer', () => {
  test('renders correctly', () => {
    const { container } = render(<TabsContainer data={mockTabData} />);
    expect(container).toMatchSnapshot();
    const tab1 = screen.getByText('Tab 1');
    const tab2 = screen.getByText('Tab 2');
    const tab3 = screen.getByText('Tab 3');
    expect(ariaSelectedValue(tab1)).toBe('true');
    expect(ariaSelectedValue(tab2)).toBe('false');
    expect(ariaSelectedValue(tab3)).toBe('false');
  });
});
