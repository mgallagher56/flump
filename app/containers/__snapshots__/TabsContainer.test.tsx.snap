// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TabsContainer > renders correctly 1`] = `
<div>
  <div
    class="chakra-tabs css-0"
    data-orientation="horizontal"
    data-part="root"
    data-scope="tabs"
    dir="ltr"
    id="tabs::r0:"
  >
    <div
      aria-orientation="horizontal"
      class="chakra-tabs__tablist css-k008qs"
      data-orientation="horizontal"
      data-part="tablist"
      data-scope="tabs"
      id="tabs::r0::list"
      role="tablist"
    >
      <button
        aria-controls="tabs::r0::content-tab1"
        aria-selected="true"
        class="chakra-button css-cf3le8"
        data-orientation="horizontal"
        data-ownedby="tabs::r0::list"
        data-part="trigger"
        data-scope="tabs"
        data-selected=""
        data-value="tab1"
        id="tabs::r0::trigger-tab1"
        role="tab"
        tabindex="0"
        type="button"
      >
        <span
          class="css-1qkvt8e"
        >
          😎
        </span>
        Tab 1
      </button>
      <button
        aria-controls="tabs::r0::content-tab2"
        aria-selected="false"
        class="chakra-button css-cf3le8"
        data-orientation="horizontal"
        data-ownedby="tabs::r0::list"
        data-part="trigger"
        data-scope="tabs"
        data-value="tab2"
        id="tabs::r0::trigger-tab2"
        role="tab"
        tabindex="1"
        type="button"
      >
        <span
          class="css-1qkvt8e"
        >
          😐
        </span>
        Tab 2
      </button>
      <button
        aria-controls="tabs::r0::content-tab3"
        aria-selected="false"
        class="chakra-button css-cf3le8"
        data-orientation="horizontal"
        data-ownedby="tabs::r0::list"
        data-part="trigger"
        data-scope="tabs"
        data-value="tab3"
        id="tabs::r0::trigger-tab3"
        role="tab"
        tabindex="2"
        type="button"
      >
        <span
          class="css-1qkvt8e"
        >
          😐
        </span>
        Tab 3
      </button>
      <div
        data-orientation="horizontal"
        data-part="indicator"
        data-scope="tabs"
        id="tabs::r0::indicator"
        style="--transition-duration: 150ms; --transition-property: left, right, top, bottom, width, height; position: absolute; will-change: var(--transition-property); transition-property: var(--transition-property); transition-duration: 0ms; transition-timing-function: var(--transition-timing-function); left: 0px; top: 0px; width: 0px; height: 0px;"
      />
    </div>
    <div
      class="chakra-tabs__tab-panels css-8atqhb"
    >
      <div
        aria-labelledby="tabs::r0::trigger-tab1"
        data-ownedby="tabs::r0::list"
        data-part="content"
        data-scope="tabs"
        data-selected=""
        id="tabs::r0::content-tab1"
        role="tabpanel"
        tabindex="0"
      >
        <div>
          Tab content 1
        </div>
      </div>
      <div
        aria-labelledby="tabs::r0::trigger-tab2"
        data-ownedby="tabs::r0::list"
        data-part="content"
        data-scope="tabs"
        hidden=""
        id="tabs::r0::content-tab2"
        role="tabpanel"
        tabindex="1"
      >
        <div>
          Tab content 2
        </div>
      </div>
      <div
        aria-labelledby="tabs::r0::trigger-tab3"
        data-ownedby="tabs::r0::list"
        data-part="content"
        data-scope="tabs"
        hidden=""
        id="tabs::r0::content-tab3"
        role="tabpanel"
        tabindex="2"
      >
        <div>
          Tab content 3
        </div>
      </div>
    </div>
  </div>
</div>
`;
